src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2069,6 +2082,252 @@ ep_rt_mono_rand_try_get_bytes (
 	return mono_rand_try_get_bytes (&_ep_rt_mono_rand_provider, (guchar *)buffer, (gssize)buffer_size, error);
 }
 
+static GString *
+quote_escape_and_append_string (char *src_str, GString *target_str)
+{
+#ifdef HOST_WIN32
+	char quote_char = '\"';
+	char escape_chars[] = "\"\\";
+#else
+	char quote_char = '\'';
+	char escape_chars[] = "\'\\";
+#endif
+
+	gboolean need_quote = FALSE;
+	gboolean need_escape = FALSE;
+
+	for (char *pos = src_str; *pos; ++pos) {
+		if (isspace (*pos))
+			need_quote = TRUE;
+		if (strchr (escape_chars, *pos))
+			need_escape = TRUE;
+	}
+
+	if (need_quote)
+		target_str = g_string_append_c (target_str, quote_char);
+
+	if (need_escape) {
+		for (char *pos = src_str; *pos; ++pos) {
+			if (strchr (escape_chars, *pos))
+				target_str = g_string_append_c (target_str, '\\');
+			target_str = g_string_append_c (target_str, *pos);
+		}
+	} else {
+		target_str = g_string_append (target_str, src_str);
+	}
+
+	if (need_quote)
+		target_str = g_string_append_c (target_str, quote_char);
+
+	return target_str;
+}
+
+char *
+ep_rt_mono_get_managed_cmd_line ()
+{
+	MONO_REQ_GC_NEUTRAL_MODE;
+
+	int argc = mono_runtime_get_main_args_argc_raw ();
+	if (argc == 0)
+		return NULL;
+
+	// managed cmdline -> native host + managed argv (which includes the entrypoint assembly)
+	size_t total_size = 0;
+	char *host_path = NULL;
+	char **argv = mono_runtime_get_main_args_argv_raw ();
+	GString *cmd_line = NULL;
+
+	host_path = minipal_getexepath();
+
+	if (host_path)
+		// quote + string + quote
+		total_size += strlen (host_path) + 2;
+
+	for (int i = 0; i < argc; ++i) {
+		if (argv [i]) {
+			if (total_size > 0) {
+				// add space
+				total_size++;
+			}
+			// quote + string + quote
+			total_size += strlen (argv [i]) + 2;
+		}
+	}
+
+	// String will grow if needed, so not over allocating
+	// to handle case of escaped characters in arguments, if
+	// that happens string will automatically grow.
+	cmd_line = g_string_sized_new (total_size + 1);
+
+	if (cmd_line) {
+		if (host_path)
+			cmd_line = quote_escape_and_append_string (host_path, cmd_line);
+
+		for (int i = 0; i < argc; ++i) {
+			if (argv [i]) {
+				if (cmd_line->len > 0) {
+					// add space
+					cmd_line = g_string_append_c (cmd_line, ' ');
+				}
+				cmd_line = quote_escape_and_append_string (argv [i], cmd_line);
+			}
+		}
+	}
+
+	g_free (host_path);
+
+	return cmd_line ? g_string_free (cmd_line, FALSE) : NULL;
+}
+
+char *
+ep_rt_mono_get_os_cmd_line ()
+{
+	MONO_REQ_GC_NEUTRAL_MODE;
+
+	// we only return the native host here since getting the full commandline is complicated and
+	// it's not super important to have the correct value since it'll only be used during startup
+	// until we have the managed commandline
+	return minipal_getexepath();
```

Comment:
EventPipe assumes that the string returned by this method can correctly be freed using g_free (os_command_line_lazy_clean). Normally that is true since we use Mono API's to create the strings that will come from the same heap as g_free will free to, but when we now switch to minipal_getexepath, that is semi true, since that method uses regular C strdup to allocate string, so in theory, that could end up allocating from one heap and free on another, but most likely they will map to the same heap. If we are going to be accurate we should either return a string that is allocated using Mono API's that should be freed using g_free or update os_command_line_lazy_clean to use free, but that will break the pattern a little.